### DISCUSSION TITLE
Future of Autokey

### DISCUSSION THREAD

**Discussion by sebastiansam55** [autokey/autokey#909] [General]
Been busy for a bit with work/life stuff. But I'll try to commit more time to the project moving forward.

I'd like to open this thread as a discussion of the best way to move forward with the project in general. Should we commit to wayland over X11? Is there any data to backup that is the way things are going? Do we know how many distros are currently shipping wayland as default over x11?

The feature set of autokey is pretty wide which makes implementing it on a new windowing server an undertaking I've underestimated 3-4 times now. 

Personally I'm of the opinion that we should get an MVP wayland release (like just the `keyboard` and `window` libraries implemented, the mouse interface is what's stopping progress currently) and just throw up a disclaimer that it's not currently 100% done. 

Thoughts? Comments? Concerns? 
@josephj11 @Elliria @BlueDrink9 @ineuw

**Comment by sebastiansam55** [comment]
Some side topics:

Qt vs GTK - maintaining two GUI frontends is a bit excessive and again adds to the development workload for pushing out changes which we should seek to minimize

Separating the Frontend from the Service - Ideally the service should be able to run independently of the frontend. I think separating this out would also reduce complexity overall. 

Long term if wayland support gets implement as I've been doing it this would also allow separation/security as we could grant a `autokey` user the input group membership needed for uinput instead of the main user of the system.

**Reply by ineuw** [reply]
Thanks for both posts, as I am sure that these matters have crossed the minds of all committed Autokey users. I feel that with our current resources, the best course of action is to update the most used GUI.

Personally, I don't care which GUI, because I am comfortable with either Qt or GTK. This philosophy also applies to X-Windows and Wayland. I would like to see a single Autokey version working seamlessly for everyone and help others to upgrade to this single version.

**Comment by Elliria** [comment]
I know how it is juggling work/life stuff with contributing to projects. We always appreciate any time you're able to give to AutoKey.
 
> Should we commit to wayland over X11?

Wayland definitely needs to happen, since at least some distributions are configuring themselves to use it by default now even though they still offer Xorg. I don't think it should happen over (or at the expense of) Xorg. Until Xorg is completely deprecated, it should remain available to those of us who choose to use it, especially since some of us have never tried Wayland and some issues with Wayland cannot yet be resolved, so folks are sometimes told to go back to Xorg for the time being.

> Is there any data to backup that is the way things are going?

The **Desktop Linux distributions** section of [this Wikipedia page](https://en.wikipedia.org/wiki/Wayland_(protocol)) lists six distributions that use Wayland by default now, and they're some big names.

Also, [this YouTube video](https://www.youtube.com/watch?v=g1BoZnekkyM) is my favorite for learning about Wayland and for seeing why it's inevitable that we must all end up using it.

> Do we know how many distros are currently shipping wayland as default over x11?

Six so far in that article linked up there, but there could be more.
 
> The feature set of autokey is pretty wide which makes implementing it on a new windowing server an undertaking I've underestimated 3-4 times now.

You're an absolute hero for taking this on. ❤️
 
> Personally I'm of the opinion that we should get an MVP wayland release (like just the `keyboard` and `window` libraries implemented, the mouse interface is what's stopping progress currently) and just throw up a disclaimer that it's not currently 100% done.

I don't fully understand that, but since you do, it sounds good. Any integration that's possible would probably be appreciated by those who use Wayland.

**Reply by josephj11** [reply]
As long as it stays an "experimental" branch, that's a good idea. The big deal would be keeping the Wayland code isolated somehow so that changes from develop can be merged into this new branch with minimal effort. We're doing too many multiple branch changes already.

**Comment by Elliria** [comment]
**Qt vs GTK:** If it comes down to a choice between them, my vote is for **GTK**, because that **Qt** editor is impossible to use.

**Reply by josephj11** [reply]
If we can swap kate-part for Qscintilla, all of that might change.

**Comment by josephj11** [comment]
I agree in principle with having one front end, but each one has a few things it does better than the other one.

We seem to  have had devs who were knowledgeable enough to work on the GTK front end, but not so much for the Qt one.

The edit panel is almost totally different between them (GTK is much better) and the panel icon works differently (Qt is better).

In previous discussions, you said there was a feature you don't use (but I do) that doesn't work on Wayland. I don't remember which one it was. I don't think it was the mouse.

I haven't gotten around to filing a feature request yet, but it would be great if a real editor could be plugged into AutoKey. Kate is built to allow that, but everyone has their own favorite editor. This would improve whichever front end we choose.

**Comment by josephj11** [comment]
> Separating the Frontend from the Service - Ideally the service should be able to run independently of the frontend. I think separating this out would also reduce complexity overall.

I believe Bluedrink9 did a bunch of refactoring that made it into 0.96.0 that started doing that. It will also remove duplicated code.

Wayland/X11 - X11 will probably need to be supported for quite a while yet because not everyone will adopt Wayland and we don't want to lose those users.

**Comment by josephj11** [comment]
For more information, see PR #844 and issue #87

**Comment by josephj11** [comment]
Based on this [discussion](https://ubuntuforums.org/showthread.php?t=2498017&p=14192125#post14192125), I think the way forward would be to rewrite [Xautomtion](https://sourceforge.net/projects/xautomation/) for Wayland (if possible) and write a fancy front end for capturing and managing tiny screenshots of features on the screen to search for. Then [ydotool](https://github.com/ReimuNotMoe/ydotool) or [dotool](https://git.sr.ht/~geb/dotool) can be used for keyboard and mouse events without us writing our own. This sounds like it would work nicely, but it would be a lot of work to design and implement.

**Comment by ineuw** [comment]
My silence was because either versions work well for me. But, my preference is GTK. As for X11 and Wayland, I support any effort to gain a wider audience for Autokey.

P.S: @josephj11, I changed my email address because the previous one was hacked. while Google One was rolled out. I also lost your email address, when several data disks were damaged unintentionally. 

Needless to say it was an expensive year. To connect with me, go to Wikipedia/Commons/Wikisource, find User:Ineuw and send me an email. The email link is accessible through one of the links on the left of my user home page. I will reply their system is secure and send you my new email address.

**Reply by josephj11** [reply]
Tried to send you  message there. Hopefully you got it.

**Reply by ineuw** [reply]
Yes. I just saw it. Thanks

**Comment by PiotrGrobelak** [comment]
> Should we commit to wayland over X11? Is there any data to backup that is the way things are going? Do we know how many distros are currently shipping wayland as default over x11?

In my work, we all use Xorg because a few tools, like Teams, don't work well with Wayland. Autokey is a very helpful tool. Used with Pano/Espano/Ulauncher, it forms a powerful automation stack for Linux. We are waiting for autumn or later to migrate to Ubuntu 24 from 22, after all applications have been prepared for Gnome 46.


In my opinion, if Ubuntu offers the option to switch between protocols, the better choice currently is Xorg. Perhaps in the future, for example with Ubuntu 26, Wayland will be more widely used and refined.

**Reply by josephj11** [reply]
Thanks for the notes. I hadn't seen pano or ulauncher before.

Maybe Wayland will work well one of these days. It's only a decade old. LOL

If your team/project relies on AutoKey, be aware that we have had no active developers for some months now.

If you know anyone with Python experience looking for a good project to work on, please send them our way. Even if we don't fix or enhance anything, we are starting to run into problems as the packages we depend on evolve.

**Reply by PiotrGrobelak** [reply]
> Maybe Wayland will work well one of these days. It's only a decade old. LOL

Perhaps AI will enhance the future of open-source. My company is small and lacks the funds to maintain that tool or offer sponsorship.

Currently, I don't have a Python developer, but I'll keep that in mind.

Personally, I've started learning Rust and might create some tools for Linux with it in the future.

**Comment by LeXofLeviafan** [comment]
Is there any release plan? I.e. what is meant to end up in the next version, and when can it be considered ready for release?

The latest published version ([0.96.0](https://pypi.org/project/autokey/#history)) is fairly old (2022-06-06) and includes a bug that affects the _builtin demo script_ (`mouse.wait_for_click()` not _actually_ listening for a mouse click), which had been fixed quite a while ago (…though only in `develop`, for some reason).

I believe a patch release at the very least is long overdue. …Preferably including that [bugfix](https://github.com/autokey/autokey/pull/636/files) :sweat_smile:

**Reply by josephj11** [reply]
Welcome to the AutoKey community, @LeXofLeviafan !  To quote individual one, I have a concept of a plan. :)

The problem is that there aren't any active developers to merge and release anything.

The former lead developer is still around for focused consultation. One other developer is still here, but doesn't appear very often and we have one junior developer who is getting better day by day, but  doesn't have the skills to make and support a new release yet. There are two new people with developer skills who are starting to help, but don't seem to want very active roles.

If you or anyone you know is interested in helping in any way, please let us know! The best place to talk things over is on [Gitter](https://gitter.im/autokey/autokeyhttps://gitter.im/autokey/autokey).
